src/common/common.h

```diff
@@ -84,21 +85,29 @@ extern FILE* g_error_fp;
   }  \
 } while(0)
 
-#define sp_debug(...) do { \
-  if (getenv("SP_DEBUG")) {   \
-  		char* nodir = basename((char*)__FILE__);				 \
-      fprintf(stderr, "%s [%d]: ", nodir, __LINE__); \
-      fprintf(stderr, __VA_ARGS__); \
-      fprintf(stderr, "\n");  \
-      fflush(stderr); \
-    } \
-  else if (getenv("SP_FDEBUG")) {               \
-  		char* nodir = basename((char*)__FILE__);				 \
-      fprintf(g_debug_fp, "%s [%d]: ", nodir, __LINE__); \
-      fprintf(g_debug_fp, __VA_ARGS__); \
-      fprintf(g_debug_fp, "\n");  \
-      fflush(g_debug_fp); \
-  }\
+#define sp_debug(debug_type, ...) do { \
+  if ((strcmp(debug_type, "injector") == 0 && getenv("SP_DEBUG_INJECTOR")) || \
+      (strcmp(debug_type, "common") == 0 && getenv("SP_DEBUG_COMMON")) || \
+      (strcmp(debug_type, "patchapi") == 0 && getenv("SP_DEBUG_PATCHAPI")) || \
+      (strcmp(debug_type, "ipc") == 0 && getenv("SP_DEBUG_IPC")) || \
+      (strcmp(debug_type, "worker") == 0 && getenv("SP_DEBUG_WORKER")) || \
+      (strcmp(debug_type, "sigtrap") == 0 && getenv("SP_DEBUG_SIGTRAP")) || \
+      (strcmp(debug_type, "agent") == 0 && getenv("SP_DEBUG_AGENT"))) { \
+    if (getenv("SP_DEBUG")) {   \
+        char* nodir = basename((char*)__FILE__);				 \
+        fprintf(stderr, "%s [%d]: ", nodir, __LINE__); \
+        fprintf(stderr, __VA_ARGS__); \
+        fprintf(stderr, "\n");  \
+        fflush(stderr); \
+      } \
+    else if (getenv("SP_FDEBUG")) {               \
+        char* nodir = basename((char*)__FILE__);				 \
+        fprintf(g_debug_fp, "%s [%d]: ", nodir, __LINE__); \
+        fprintf(g_debug_fp, __VA_ARGS__); \
+        fprintf(g_debug_fp, "\n");  \
+        fflush(g_debug_fp); \
+    }\
+  } \
```

Comment:
set `g_debug_fp` to `stderr` if `SP_DEBUG` environment variable is set, or `FILE*` returned by `fopen` if `SP_FDEBUG` is environment variable is set.  Only set `debugTypeEnabled` values is either is set, and only `fclose` file is `g_debug_fp` is not `nullptr` or `stderr`.